Add ability to set a classical condition on a list of bits

enavarro51 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Use list of classical bits in conditions

### Details and comments

This PR is a continuation of #6761 by @TharrmashasthaPV. That PR created all the underlying code to represent conditions as a list of bits. This PR extends this by modifying the 3 circuit drawers to properly display conditions that are lists. For example,
```
clbits = [Clbit(), Clbit(), Clbit()]
qr = QuantumRegister(2, "qr")
cr = ClassicalRegister(3, "cr")
circuit = QuantumCircuit(qr, clbits, cr)
circuit.x(0).c_if([cr[0], clbits[1], cr[1], clbits[2]], 11)
circuit.draw("mpl", cregbundle=False)
```
![image](https://user-images.githubusercontent.com/16268251/153665513-6b9232ab-658f-4c95-94d7-1626c29c8069.png)
As part of this, if `cregbundle` is on and at least one of the bits in the list belongs to a register, then `cregbundle` is turned off with a warning to the user. This allows full display of the details of the condition.

Also now all conditions in the 'mpl' drawer display the hex value at the bottom to be consistent.

javabster commented on qiskit/visualization/matplotlib.py:
```diff
@@ -406,11 +406,29 @@ def _get_layer_widths(self):
                         RuntimeWarning,
                         2,
                     )
+                if (
```
This snippet looks very similar to what's in `latex.py`, would it make sense to move this out to a utils file and reduce duplication of logic?

enavarro51 replied:
Actually basically matches the 'text' drawer code as well, though it handles the `cargs` version differently. Could definitely be added to utils. Will look into it.

enavarro51 replied:
@javabster Moved the `cregbundle` checks to `utils.py` in https://github.com/Qiskit/qiskit-terra/pull/7653/commits/829ded3621a87d63b7e4dca8fb0aa180f65e15c1.

javabster commented on qiskit/visualization/utils.py:
```diff
@@ -281,40 +281,52 @@ def get_wire_label(drawer, register, index, layout=None, cregbundle=True):
 
 
-def get_condition_label_val(condition, circuit, cregbundle, reverse_bits):
+def get_condition_label_val(condition, circuit, cregbundle):
```
Maybe beyond the scope of this PR, but I think as this utils file keeps growing it would be good at some point to separate these helper functions out into multiple files, in their own utils folder

javabster replied:
That would be great! And it would be great to get an architecture diagram put together to go with with that, I feel like the visualisation code can be quite confusing for newcomers so something like that would help a lot

javabster replied:
again, beyond the scope of this PR but food for thought in future 😄

enavarro51 commented:
@javabster The `cregbundle `checks are now in `utils`. The mpl drawer goes through the nodes before drawing in order to get layer widths. I added the checks into this for loop. The latex drawer also goes through the nodes at the beginning to get `image_width`, and I added the checks there. The text drawer does not have a pre-processing loop, so I had to make an extra trip through the nodes to do the `cregbundle `check. Should be ready to go now.

javabster reviewed (dismissed):
This LGTM, thanks Ed! I think @mtreinish might have some thoughts on the dagcircuit changes?

1ucian0 commented on qiskit/circuit/instruction.py:
```diff
@@ -405,17 +405,30 @@ def inverse(self):
+            raise CircuitError("condition value should be less than 2 ^ number of bits")
+
+        if isinstance(classical, Clbit) and int(val) > 1:
+            raise CircuitError("condition value should be 0/1 or True/False")
```
I tried to trigger this exception (in order to create a test for it). But it seems like it is covered by line 421 
> raise CircuitError("condition value should be non-negative")

1ucian0 replied:
Triggered! testing in 9cafc8571

Still not fully sure how 421 does not get this one.

1ucian0 commented on qiskit/circuit/instruction.py:
```diff
@@ -405,17 +405,30 @@ def inverse(self):
+        if (isinstance(classical, ClassicalRegister) and val >= 2**classical.size) or (
+            isinstance(classical, list) and val >= 2 ** len(classical)
+        ):
+            raise CircuitError("condition value should be less than 2 ^ number of bits")
```
tested in 9a29cfc31e990cedc4f8e060165b7a7caee16418

1ucian0 commented on qiskit/visualization/text.py:
```diff
@@ -676,6 +677,17 @@ def __init__(
+
+        if self.cregbundle:
+            for layer in nodes:
+                for node in layer:
+                    if self.cregbundle:
+                        self.cregbundle = check_cregbundle(node, self._circuit)
+                    else:
+                        break
+                if not self.cregbundle:
+                    break
```
These nested loops are a bit hard to follow. I think it is  a case of [for/else](https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops)

1ucian0 reviewed (approved):
Thanks @enavarro51 ! I added some error-handling tests for completeness. Let's merge it!

jakelishman commented:
I've been away for quite a while, so never quite got time to comment on this before it merged.  Sorry that this is coming a bit out of nowhere now.

This PR has caused relatively serious regressions in several conversion functions, though those are at least overshadowed by other factors over a complete transpilation workload.  Having the visualisations prepared to work with conditions on several bits is certainly very useful, but the problem is that this idea of conditioning on a list of bits like this is not how we'll be moving forwards with the more complex conditionals used in dynamic circuits.  This PR only adds support for this style of condition to the visualisations, but the chance of it being added to any simulator is very low, and certainly won't be in the 0.20 release.

I think it might be best to roll back this PR, rather than release known regressions into 0.20 for a feature that doesn't have a use yet.  The concept of "conditioning on a list of bits" will exist in the future, and I think a large amount of this PR will be useful (especially the visualisation components), but I think it's best not to extend the `(bit, value)` 2-tuple when it's something that we'll want to be deprecating.  Simulator support for conditionals on lists of bits won't be available until the classical type system is in place, so I think it may be better to hold off on this until we're ready to do it a bit more completely.

jakelishman commented:
This idea of "conditioning on a list of bits" is actually already supported by Terra, it's just not supported on ad-hoc lists like this PR does.  You can currently just add a register to the circuit that contains bits from other registers, and apply the condition on that.

Perhaps the best course of action is to roll back the changes to `Instruction` that attempt to add list of bits conditionals, and keep the visualisation components.  We could update the visualisation components so they work when applied to a register that is not one of the drawn ones, or one that is not defined over contiguous clbits.

The changes to `Instruction.condition` here are a bit inconsistent, as well.  You can condition on a single bit by integer or instance (`c_if(0)` or `c_if(bits[0])`), but you can only condition on a list of bit instances:
```python

jakelishman commented:
To be clear (and sorry for the third long message): I think we should roll back this PR, but then use the new visualiation logic to make this (currently erroring) code work:
```python
from qiskit.circuit import QuantumCircuit, ClassicalRegister, Clbit, QuantumRegister

clbits = [Clbit() for _ in [None] * 5]
cr1 = ClassicalRegister(bits=clbits)
cr2 = ClassicalRegister(bits=clbits[::2])
qc = QuantumCircuit(QuantumRegister(5), cr1, cr2)
for i in range(5):
    qc.h(i)
qc.measure(range(5), range(5))
qc.x(0).c_if(cr2, 2**3 - 1)
qc.draw()  # ValueError in Layer.set_clbit due to failed register lookup
```
This shows how "lists of bits" conditions could already be emulated (just not in an ad-hoc manner, and still with no simulator support), and the new visualisation changes I think can be used to make this work in the drawers, which is very useful.

chriseclectic commented:
@jakelishman This PR is also causing test failures on qiskit-experiments, I confirmed reverting this commit fixed the failing tests.
